lib/AST/Type.cpp

```diff
@@ -3976,10 +3976,13 @@ operator()(CanType dependentType, Type conformingReplacementType,
           || conformingReplacementType->is<DependentMemberType>()
           || conformingReplacementType->is<TypeVariableType>())
          && "replacement requires looking up a concrete conformance");
-  // Lookup conformances for opened existential.
-  if (conformingReplacementType->isOpenedExistential()) {
-    return conformedProtocol->getModuleContext()->lookupConformance(
-        conformingReplacementType, conformedProtocol);
+  // A class-constrained archetype might conform to the protocol
+  // concretely.
+  if (auto *archetypeType = conformingReplacementType->getAs<ArchetypeType>()) {
+    if (auto superclassType = archetypeType->getSuperclass()) {
+      return conformedProtocol->getModuleContext()->lookupConformance(
+          archetypeType, conformedProtocol);
+    }
```

Comment:
Should `MakeAbstractConformanceForGenericType` perhaps consistently fail when passed an archetype and the protocol is not listed in its conformances (not just with class-constrained ones)?

Reply:
Are you suggesting this as an assertion?

Reply:
Not quite — that we consistently lookupConformance() for all archetypes, or else an invalid conformance will be returned only for *class-constrained* archetypes that don't conform.

Reply:
It might be best to replace all usages of MakeAbstractConformanceForGenericType with LookUpConformanceInModule, in fact. There's no harm in using the latter on abstract archetypes or generic parameters. However I want to keep this change relatively narrow since it unblocks some work I'm doing internally.